### ISSUE TITLE
Runtime error with tuples

### ISSUE THREAD

**Issue by TheSeamau5** [elm-lang/core#201]
I am unable to pin down why this error happens but the failing code is surprisingly short:

```elm
smap : (a -> b -> (a,b)) -> a -> List b -> (a, List b)
smap f a list = case list of
  [] -> (a, [])
  b :: bs  ->
    let (a',b') = f a b
        (a'', list) = smap f a' bs
    in
        (a'', b' :: list)

test = smap (,) 0 [0]
```

which causes the following error:

```
Cannot read property '_0' of undefined


Open the developer console for more details.
```

**Comment by jvoigtlaender** [comment]
I think this is a compiler error that has to do with name shadowing. Note that the runtime error goes away if you change the code to this:
```elm
smap : (a -> b -> (a,b)) -> a -> List b -> (a, List b)
smap f a list = case list of
  [] -> (a, [])
  b :: bs  ->
    let (a',b') = f a b
        (a'', list') = smap f a' bs
    in
        (a'', b' :: list)
```
or to this:
```elm
smap : (a -> b -> (a,b)) -> a -> List b -> (a, List b)
smap f a list = case list of
  [] -> (a, [])
  b :: bs  ->
    let (a',b') = f a b
        (a'', list') = smap f a' bs
    in
        (a'', b' :: list')
```
Can you close this issue here, check whether a corresponding issue is already open at the `elm-compiler` repo, and if not, open an issue there?

**Comment by evancz** [comment]
When you reopen, do you mind also putting in the generated JS? I think that'd help a lot in figuring out exactly what went wrong.

**Comment by rtfeldman** [comment]
Cannot reproduce in 0.16.

I pasted this into try-elm:

```elm
import Html exposing (text)

main =
  text ("Hello, World!" ++ toString (test))
  
smap : (a -> b -> (a,b)) -> a -> List b -> (a, List b)
smap f a list = case list of
  [] -> (a, [])
  b :: bs  ->
    let (a',b') = f a b
        (a'', list) = smap f a' bs
    in
        (a'', b' :: list)

test = smap (,) 0 [0]
```

...and the screen displayed:

```
Hello, World!(0,[0])
```

I think this may have been fixed by the codegen updates in 0.16.

**Comment by evancz** [comment]
Cool, thanks for following up on this!
